patches/api/0377-Add-Position.patch

```diff
@@ -0,0 +1,382 @@
++     * @return the offset position
++     */
++    @NotNull FinePosition offset(double x, double y, double z);
++
++    /**
++     * Gets a position for the block above.
++     *
++     * @return a position for the block above
++     */
++    @Contract("-> new")
++    default @NotNull BlockPosition blockAbove() {
++        return this.blockAbove(1);
++    }
++
++    /**
++     * Gets a position for the block the specified distance above
++     *
++     * @param distance the distance to travel up
++     * @return a position for the block
++     */
++    default @NotNull BlockPosition blockAbove(int distance) {
++        return distance == 0 && this instanceof BlockPosition blockPos ? blockPos : new BlockPositionImpl(this.blockX(), this.blockY() + distance, this.blockZ());
++    }
++
++    /**
++     * Gets a position for the block below.
++     *
++     * @return a position for the block below.
++     */
++    @Contract("-> new")
++    default @NotNull BlockPosition blockBelow() {
++        return this.blockBelow(1);
++    }
++
++    /**
++     * Gets a position for the block the specified distance below
++     *
++     * @param distance the distance to travel down
++     * @return a position for the block
++     */
++    default @NotNull BlockPosition blockBelow(int distance) {
++        return distance == 0 && this instanceof BlockPosition blockPos ? blockPos : new BlockPositionImpl(this.blockX(), this.blockY() - distance, this.blockZ());
++    }
++
++    /**
++     * Gets a position offset up by the specified distance.
++     *
++     * @param distance the distance to travel up
++     * @return the offset position
++     */
++    @Contract("_ -> new")
++    default @NotNull FinePosition above(double distance) {
++        return new FinePositionImpl(this.x(), this.y() + distance, this.z());
++    }
++
++    /**
++     * Gets a position offset down by the specified distance.
++     *
++     * @param distance the distance to travel down
++     * @return the offset position
++     */
++    @Contract("_ -> new")
++    default @NotNull FinePosition below(double distance) {
++        return new FinePositionImpl(this.x(), this.y() - distance, this.z());
++    }
++
++    /**
++     * Returns a new position at the center of the block position this represents
++     *
++     * @return a new center position
++     */
++    @Contract("-> new")
++    default @NotNull FinePosition toCenter() {
++        return new FinePositionImpl(this.blockX() + 0.5, this.blockY() + 0.5, this.blockZ() + 0.5);
++    }
++
++    /**
++     * Returns the block position of this position
++     * or itself if it already is a block position
++     *
++     * @return the block position
++     */
++    default @NotNull BlockPosition toBlock() {
++        return new BlockPositionImpl(this.blockX(), this.blockY(), this.blockZ());
++    }
++
++    /**
++     * Converts this position to a vector
++     *
++     * @return a new vector
++     */
++    @Contract("-> new")
```

Comment:
```suggestion
+    @Contract(value = "-> new", pure = true)
```

Reply:
maybe it's time for contract-validator? :p